module/surface_sam_sid_vhf.c

```diff
@@ -0,0 +1,411 @@
+{
+	hid_dbg(hid, "%s\n", __func__);
+	return 0;
+}
+
+static void sid_vhf_hid_stop(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+}
+
+static int sid_vhf_hid_open(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+	return 0;
+}
+
+static void sid_vhf_hid_close(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+}
+
+struct surface_sam_sid_vhf_meta_rqst {
+	u8 id;
+	u32 offset;
+	u32 limit;
+	u8 end; // 0x01 if end was reached
+} __packed;
+
+struct vhf_meta_info_resp {
+	u8 len;
+	u8 _2;
+	u8 _3;
+	u8 _4;
+	u8 _5;
+	u8 _6;
+	u8 _7;
+	u16 hid_len; // hid descriptor length
+} __packed;
+
+struct vhf_meta_meta_resp {
+    u32 len;
+    u16 vendor_id;
+    u16 product_id;
+    u8  _1[24];
+} __packed;
+
+union vhf_buffer_data {
+	struct vhf_meta_info_resp info;
+	u8 pld[0x76];
+	struct vhf_meta_meta_resp meta;
+};
+
+struct surface_sam_sid_vhf_meta_resp {
+	struct surface_sam_sid_vhf_meta_rqst rqst;
+	union vhf_buffer_data data;
+} __packed;
+
+
+static int vhf_get_metadata(u8 iid, struct vhf_meta_meta_resp *meta)
+{
+	int status;
+	int len;
+	u8 *buf;
+
+	struct surface_sam_sid_vhf_meta_resp *resp;
+	struct surface_sam_ssh_rqst rqst = {};
+	struct surface_sam_ssh_buf result = {};
+
+	resp = kzalloc(sizeof(struct surface_sam_sid_vhf_meta_resp), GFP_KERNEL);
+
+	resp.rqst.id = 2;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	rqst.tc = 0x15;
+	rqst.cid = 0x04;
+	rqst.iid = iid;
+	rqst.pri = 0x02;
+	rqst.snc = 0x01;
+	rqst.cdl = sizeof(struct surface_sam_sid_vhf_meta_rqst);
+	rqst.pld = (u8*)&resp.rqst;
+
+	result.cap  = sizeof(struct surface_sam_sid_vhf_meta_resp);
+	result.len  = 0;
+	result.data = (u8*)resp;
+
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	if (status) {
+		return status;
+	}
+
+	*meta = resp;
+}
+
+static int vhf_get_hid_descriptor(struct hid_device *hid, u8 iid, u8 **desc, int *size)
+{
+	int status;
+	int len;
+	u8 *buf;
+
+	struct surface_sam_ssh_rqst rqst = {};
+	struct surface_sam_sid_vhf_meta_resp resp = {};
+	struct surface_sam_ssh_buf result = {};
+
+	resp.rqst.id = 0;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	rqst.tc = 0x15;
+	rqst.cid = 0x04;
+	rqst.iid = iid;
+	rqst.pri = 0x02;
+	rqst.snc = 0x01;
+	rqst.cdl = sizeof(struct surface_sam_sid_vhf_meta_rqst);
+	rqst.pld = (u8*)&resp.rqst;
+
+	result.cap  = sizeof(struct surface_sam_sid_vhf_meta_resp);
+	result.len  = 0;
+	result.data = (u8*)&resp;
+
+	// first fetch 01 to get the total length
+	status = surface_sam_ssh_rqst(&rqst, &result);
+	if (status) {
+		return status;
+	}
+
+	len = resp.data.info.hid_len;
+
+	// allocate a buffer for the descriptor
+	buf = kzalloc(len, GFP_KERNEL);
+
+	// then, iterate and write into buffer, copying out resp.rqst.hid_len bytes
+	// (it gets set to the length of data read)
+
+	resp.rqst.id = 1;
+	resp.rqst.offset = 0;
+	resp.rqst.limit = 0x76;
+	resp.rqst.end = 0;
+
+	while (resp.rqst.end != 0x01 && resp.rqst.offset < len) {
```

Comment:
It may be a bit better to treat `end` here as boolean value, so just `while (!resp.rqst.end && ...` as we don't know if that could ever take on some other values.